core/cache/namespace.go

```diff
@@ -0,0 +1,79 @@
+package cache
+
+import (
+	"context"
+	"time"
+
+	"github.com/go-logr/logr"
+	v1 "k8s.io/api/core/v1"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+const pollIntervalSeconds = 120
+
+type namespaceStore struct {
+	client       client.Client
+	namespaces   []v1.Namespace
+	logger       logr.Logger
+	forceRefresh chan bool
+	cancel       func()
+}
+
+func newNamespaceStore(c client.Client, logger logr.Logger) namespaceStore {
+	return namespaceStore{
+		client:       c,
+		namespaces:   []v1.Namespace{},
+		logger:       logger,
+		cancel:       nil,
+		forceRefresh: make(chan bool),
+	}
+}
+
+func (n *namespaceStore) Namespaces() []v1.Namespace {
+	return n.namespaces
+}
+
+func (n *namespaceStore) ForceRefresh() {
+	n.forceRefresh <- true
+}
+
+func (n *namespaceStore) Stop() {
+	if n.cancel != nil {
+		n.cancel()
+	}
+}
+
+func (n *namespaceStore) Start(ctx context.Context) {
+	var newCtx context.Context
+
+	newCtx, n.cancel = context.WithCancel(ctx)
+
+	go func() {
+		ticker := time.NewTicker(pollIntervalSeconds * time.Second)
+
+		defer ticker.Stop()
+
+		for {
+			list := &v1.NamespaceList{}
+
+			err := n.client.List(newCtx, list)
```

Comment:
Question: so once we switch to the client pool, this will be a loop over all the clients to populate the namespace cache, correct?

Reply:
There may need to be a little rethink of how we interleave these 2 things together.

Right now the ClientPool is build ad-hoc per request (see top func of core/server/kustomization.go), whereas this cache is started (with a client) on server boot.

2 things may need to happen:
- It would be helpful if the ClientPool was cached in advance, not built on the fly (I believe we have plans for this)
- This Start may either need to watch the context for the client pool key to appear, or we move _something_ up to where the query is happening


My hot take, off for a covid test

Reply:
> Question: so once we switch to the client pool, this will be a loop over all the clients to populate the namespace cache, correct?

It can loop over them, but it is possible to make it parallel and each client in the pool has it's own cycle, only `Start()` and `Stop()` would be exposed to do that. I don't really know how client pool is working 100% but I tried to make this logic isolated as possible so we can change anything inside and all callers would still use the `Namespaces()` function, that might be better to change to return `map[string]v1.Namespace` instead.

Reply:
Good point, right now `CientsPool` expects a user to impersonate, but in this context, there is no user, since it's the server calling with its own credentials. We could tweak a little bit to ignore the impersonation if the user is `nil` to make that happen.

The cache container will need to have its own Clients Pool built once on cache container instantiation.